controllers/loadtest_controller.go

```diff
@@ -175,131 +176,221 @@ func (r *LoadTestReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+			return ctrl.Result{Requeue: true}, errCacheSync
 		}
 
-		if err = r.Create(ctx, pod); err != nil {
-			log.Error(err, "could not create new pod", "pod", pod)
-			return &ctrl.Result{Requeue: true}, err
+		nodes := new(corev1.NodeList)
+		if err = r.List(ctx, nodes); err != nil {
+			log.Error(err, "failed to list nodes")
+			return ctrl.Result{Requeue: true}, err
 		}
 
-		return nil, nil
-	}
-
-	for i := range missingPods.Servers {
-		logWithServer := log.WithValues("server", missingPods.Servers[i])
-
-		pod, err := builder.PodForServer(&missingPods.Servers[i])
-		if err != nil {
-			logWithServer.Error(err, "failed to construct a pod struct for supplied server struct")
-
-			test.Status.State = grpcv1.Errored
-			test.Status.Reason = grpcv1.ConfigurationError
-			test.Status.Message = fmt.Sprintf("failed to construct a pod for server at index %d: %v", i, err)
+		var defaultClientPool string
+		var defaultDriverPool string
+		var defaultServerPool string
+		poolCapacities := make(map[string]int)
+		for _, node := range nodes.Items {
+			pool, ok := node.Labels[config.PoolLabel]
+			if !ok {
+				log.Info("encountered a node without a pool label", "nodeName", node.Name)
+				continue
+			}
 
-			if updateErr := r.Status().Update(ctx, test); updateErr != nil {
-				logWithServer.Error(updateErr, "failed to update status after failure to construct a pod for server")
+			if defaultPoolLabels := r.Defaults.DefaultPoolLabels; defaultPoolLabels != nil {
+				if defaultClientPool == "" {
+					if _, ok := node.Labels[defaultPoolLabels.Client]; ok {
+						defaultClientPool = pool
+					}
+				}
+				if defaultDriverPool == "" {
+					if _, ok := node.Labels[defaultPoolLabels.Driver]; ok {
+						defaultDriverPool = pool
+					}
+				}
+				if defaultServerPool == "" {
+					if _, ok := node.Labels[defaultPoolLabels.Server]; ok {
+						defaultServerPool = pool
+					}
+				}
+
+				if _, ok = poolCapacities[pool]; !ok {
+					poolCapacities[pool] = 0
+				}
 			}
 
-			return ctrl.Result{}, err
+			poolCapacities[pool]++
 		}
 
-		result, err := createPod(pod)
-		if result != nil && !kerrors.IsAlreadyExists(err) {
-			logWithServer.Error(err, "failed to create pod for server")
-
-			test.Status.State = grpcv1.Errored
-			test.Status.Reason = grpcv1.KubernetesError
-			test.Status.Message = fmt.Sprintf("failed to create pod for server at index %d: %v", i, err)
-
-			if updateErr := r.Status().Update(ctx, test); updateErr != nil {
-				logWithServer.Error(updateErr, "failed to update status after failure to create pod for server")
+		poolAvailabilities := make(map[string]int)
+		for pool, capacity := range poolCapacities {
+			poolAvailabilities[pool] = capacity
+		}
+		for _, pod := range pods.Items {
+			pool, ok := pod.Labels[config.PoolLabel]
+			if !ok {
+				log.Info("encountered a pod without a pool label", "pod", pod)
+				continue
+			}
+			if pod.Status.Phase != corev1.PodSucceeded && pod.Status.Phase != corev1.PodFailed {
```

Comment:
Wondering why we used pod.Status.Phase here, as you have advised me earlier it might not be fully updated.

Reply:
Good question. Let me remember what the issue was. I believe it had to do with Pending or Initializing pods. I do not think the status.StateForPod function worked since they had no containers. I'll recheck and update here.